packages/cardpay-sdk/sdk/claim-settlement-module.ts

```diff
@@ -19,39 +27,265 @@ export default class ClaimSettlementModule extends SafeModule {
+    avatarAddress?: string,
+    validatorAddress?: string,
+    txnOptions?: TransactionOptions,
+    contractOptions?: ContractOptions
+  ): Promise<SuccessfulTransactionReceipt> {
+    //TODO: Multi-signature. Only supports adding single validator only
+    if (!avatarAddress) {
+      throw new Error('avatarAddress must be specified');
+    }
+    if (!validatorAddress) {
+      throw new Error('validatorAddress must be specified');
+    }
+    let { nonce, onNonce, onTxnHash } = txnOptions ?? {};
+    let module = new Contract(moduleAddress, this.abi, this.ethersProvider);
+    let data = await module.interface.encodeFunctionData('addValidator', [validatorAddress]);
+    let signer = this.signer ? this.signer : this.ethersProvider.getSigner();
+    let from = contractOptions?.from ?? (await signer.getAddress());
+
+    let safe = new Contract(avatarAddress, GnosisSafeABI, this.ethersProvider);
+    const safeOwners = await safe.methods.getOwners();
+    if (!safeOwners.contain(signer)) {
+      throw new Error(`${signer} is not owner of avatar ${avatarAddress}`);
+    }
+
+    let estimate = await gasEstimate(
+      this.ethersProvider,
+      avatarAddress,
+      module.address,
+      '0',
+      data,
+      Operation.CALL,
+      AddressZero,
+      true
+    );
+
+    if (nonce == null) {
+      nonce = getNextNonceFromEstimate(estimate);
+      if (typeof onNonce === 'function') {
+        onNonce(nonce);
+      }
+    }
+
+    let gnosisTxn = await executeTransaction(
+      this.ethersProvider,
+      avatarAddress,
+      module.address,
+      data,
+      Operation.CALL,
+      estimate,
+      nonce,
+      await signSafeTx(
+        this.ethersProvider,
+        avatarAddress,
+        module.address,
+        data,
+        Operation.CALL,
+        estimate,
+        nonce,
+        from,
+        this.signer
+      )
+    );
+
+    let txnHash = gnosisTxn.ethereumTx.txHash;
+    if (typeof onTxnHash === 'function') {
+      await onTxnHash(txnHash);
+    }
+    return await waitUntilTransactionMined(this.ethersProvider, txnHash);
+  }
+
+  async defaultClaim(moduleAddress: string, payeeAddress: string) {
+    let id = utils.hexlify(utils.randomBytes(32));
+    let startBlockNum = await this.ethersProvider.getBlockNumber();
+    let startBlockTime = (await this.ethersProvider.getBlock(startBlockNum)).timestamp;
+    let validitySeconds = 86400; //1 day
+    let tokenAddress = '0x95093b8836ED53B4594EC748995E45b0Cd2b1389'; // CTST goerli
+    let transferAmount = BigNumber.from(utils.parseUnits('1', 'ether'));
+    return new Claim(
+      id,
+      this.ethersProvider.network.chainId.toString(),
+      moduleAddress,
+      new TimeRangeSeconds(startBlockTime, startBlockTime + validitySeconds),
+      new Address(payeeAddress),
+      new TransferERC20ToCaller(tokenAddress, transferAmount)
+    );
+  }
+
+  async checkValidity(claim: Claim, moduleAddress: string, callerAddress: string): Promise<void> {
```

Comment:
Can we change this to do a single call and check the revert reason?

Reply:
Fixed here https://github.com/cardstack/cardstack/pull/3641/commits/450cc44b942aaff52bb4a2aa0b7e48f54678b8f4. Basically we can use a `staticCall` to test out the transaction. It doesn't cover cases such as "not having gas to pay for the transaction", but I believe it removes the calls made inside `checkValidity` and `isValidator`. 

The call will throw an exception `CALL_EXCEPTION` such as below

```
Error: call revert exception; VM Exception while processing transaction: reverted with reason string "Caller cannot claim" [ See: https://links.ethers.org/v5-errors-CALL_EXCEPTION ] (method="signedExecute(bytes,bytes)", data="0x08c379a00000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000001343616c6c65722063616e6e6f7420636c61696d00000000000000000000000000", errorArgs=["Caller cannot claim"], errorName="Error", errorSignature="Error(string)", reason="Caller cannot claim", code=CALL_EXCEPTION, version=abi/5.7.0)
    at Logger.makeError (/Users/tintinthong/Github/cardstack/node_modules/@ethersproject/logger/src.ts/index.ts:269:28)
    at Logger.throwError (/Users/tintinthong/Github/cardstack/node_modules/@ethersproject/logger/src.ts/index.ts:281:20)
    at Interface.decodeFunctionResult (/Users/tintinthong/Github/cardstack/node_modules/@ethersproject/contracts/node_modules/@ethersproject/abi/src.ts/interface.ts:427:23)
    at Object.<anonymous> (/Users/tintinthong/Github/cardstack/node_modules/@ethersproject/contracts/src.ts/index.ts:400:44)
    at step (/Users/tintinthong/Github/cardstack/node_modules/@ethersproject/contracts/lib/index.js:48:23)
    at Object.next (/Users/tintinthong/Github/cardstack/node_modules/@ethersproject/contracts/lib/index.js:29:53)
    at fulfilled (/Users/tintinthong/Github/cardstack/node_modules/@ethersproject/contracts/lib/index.js:20:58)
    at processTicksAndRejections (internal/process/task_queues.js:95:5) {
  reason: 'Caller cannot claim',
  code: 'CALL_EXCEPTION',
  method: 'signedExecute(bytes,bytes)',
  data: '0x08c379a00000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000001343616c6c65722063616e6e6f7420636c61696d00000000000000000000000000',
  errorArgs: [ 'Caller cannot claim' ],
  errorName: 'Error',
  errorSignature: 'Error(string)',
  address: '0xcF278C499C04F8755591D73d6c49589a203b3a44',
  args: [
    '0xf474e2e9c0d28b3e6fb4d228a32cd3aa555d6d5b0e71e449a1135c1ef3eb3c2e60a88bf4954f158d779c12abc5799b61610ce225c17094efcb31707460c876a21c',
    '0x675f2ffd4fc99a4b7fd872754d46f98b1317ac40e61b62e830f0b1b21c550d5428740ee37c56efc894d649ac7cfae9bf1e56c923066eb6ff0bd5cf5e5d1486f304cf0c5b1db4f67bcd80f58b8278f67770c2510d0ff802430de4c4df8418eda30000000000000000000000000000000000000000000000000000000000000120b8fbceb1d7fb3cc75b8afdb373227d27e0d4499fedad5b62b12b9c6f8ea7de930000000000000000000000000000000000000000000000000000000000000180d2f580d93fd63852846304057772a7bca9c7d7e65a76a8ab33831ffb4c91323300000000000000000000000000000000000000000000000000000000000001c0000000000000000000000000000000000000000000000000000000000000022000000000000000000000000000000000000000000000000000000000000000400000000000000000000000000000000000000000000000000000000063f84e580000000000000000000000000000000000000000000000000000000063f99fd800000000000000000000000000000000000000000000000000000000000000200000000000000000000000000110d49df753b293d85c8a2a3820edfdfc178ee8000000000000000000000000000000000000000000000000000000000000004000000000000000000000000095093b8836ed53b4594ec748995e45b0cd2b13890000000000000000000000000000000000000000000000000de0b6b3a76400000000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000016345785d8a0000'
  ],
  transaction: {
    data: '0xd42ede80000000000000000000000000000000000000000000000000000000000000004000000000000000000000000000000000000000000000000000000000000000c00000000000000000000000000000000000000000000000000000000000000041f474e2e9c0d28b3e6fb4d228a32cd3aa555d6d5b0e71e449a1135c1ef3eb3c2e60a88bf4954f158d779c12abc5799b61610ce225c17094efcb31707460c876a21c000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000260675f2ffd4fc99a4b7fd872754d46f98b1317ac40e61b62e830f0b1b21c550d5428740ee37c56efc894d649ac7cfae9bf1e56c923066eb6ff0bd5cf5e5d1486f304cf0c5b1db4f67bcd80f58b8278f67770c2510d0ff802430de4c4df8418eda30000000000000000000000000000000000000000000000000000000000000120b8fbceb1d7fb3cc75b8afdb373227d27e0d4499fedad5b62b12b9c6f8ea7de930000000000000000000000000000000000000000000000000000000000000180d2f580d93fd63852846304057772a7bca9c7d7e65a76a8ab33831ffb4c91323300000000000000000000000000000000000000000000000000000000000001c0000000000000000000000000000000000000000000000000000000000000022000000000000000000000000000000000000000000000000000000000000000400000000000000000000000000000000000000000000000000000000063f84e580000000000000000000000000000000000000000000000000000000063f99fd800000000000000000000000000000000000000000000000000000000000000200000000000000000000000000110d49df753b293d85c8a2a3820edfdfc178ee8000000000000000000000000000000000000000000000000000000000000004000000000000000000000000095093b8836ed53b4594ec748995e45b0cd2b13890000000000000000000000000000000000000000000000000de0b6b3a76400000000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000016345785d8a0000',
    to: '0xcF278C499C04F8755591D73d6c49589a203b3a44'
  }
}
```

Reply:
I was not sure if I wanted to capture exception and rethrow. But given this is a generic [CALL_EXCEPTION](https://docs.ethers.org/v5/single-page/#/v5/api/utils/logger/-%23-errors--call-exception), I figured ill just leave it to throw as it does.